### DISCUSSION TITLE
Some feature suggestions

### DISCUSSION THREAD

**Discussion by AlexOnGitHubb** [winstxnhdw/lc-hax#508] [Ideas]
Hello LC Hax team.

Your feature rich client is awesome. Thank you so much for the time and effort you put into developing the features.
Here are some ideas I think will make nice editions/changes to the hacked client.

**Full disclosure I have no coding experience and have no clue if any of this is possible. I do hope you will consider/investigate the possibility of my suggestions.**
Note: The most recent build I used is the March 2nd 2024 build (or at least that's when I used it last. I mainly used LC Hax during Jan and Feb 2024. So a lot of my suggestions/QoL requests are based on that time.

I apologise if some of my suggestions have already been implemented.

- **Change M3** click on player to **lure just one** random entity/enemy to the target player and **/hate lures all.**
I like doing light trolling with my friends. Sending all enemies to one player is very obvious especially if its a spider or snare flea. I think M3 click on a target only lures one enemy and the `/hate` command lured all enemies. It extends my fun watching them fight/run plus gives more variety in use. You could also have the command `/hate` be used like this `/hate <player> <entity>` to specifically lure (or the command can be `/lure`) an enemy to a player.

- **Instant Easter Egg Explosion / All Eggs explode**
As part of v50 easter eggs are a scrap item that randomly explode when dropped. Like with `/explode` for jetpacks extending the command with `/explode egg` to force all eggs on ground or in player/entity's hands (or inventory if possible) to instantly explode.
Additionally or alternatively any egg the user picks up, when thrown, explodes on impact every time.

- **Tulip Snake on player**
Also as part of v50, Tulip Snakes are like mantacoils and snare flees. They occasionally target a player and land on their head and slightly obstruct view. This could be an extension of the `/hate` (or `/lure`) suggestion above where a targeted player has a snake sent to them and obstructs their view. A stand alone command for this could just be `/snake <player>`. This should work regardless if target is inside or outside.

- **Speed and Jump Modifiers.**
There is a glitch in LC that allows the player to have negative weight which makes them run really fast. Using this or some other method to modify the player/user's weight and allow them to move quickly. This could be used in a command like `/Speed <value>` or `/speed <player> <value>` if that's possible.
Additionally this would be fun to also have as a jump modifier too. It would work the same way. I am aware of `/Jump` giving infinite jumps.

- **No Drowning**
A simple feature that allows the user to never drown. This can be a patch or command, doesn't matter to me.
An extension of this could be the ability to never lose stamina underwater, then we can jump freely underwater.

- **Force Spike Trap to Slam**
v50 introduces a new indoor trap, spike traps. These either slam down at set intervals or when a player walks under it. M3 clicking on the trap could instantly activate it to slam regardless of its timer or if there is a player under it. This could be a command that activates all, doesn't matter much to me. `/Spike`

- **No Restrictions on Furniture Placement**
Furniture is locked to only be inside the ship. This could be changed to allow them to be placed outside the ship and also at any Y-level (e.g. Red Cabinet is not locked to floor). This could be a command but would probably make more sense as a mod.

- **Disco Command**
`/Disco` Places the (v50) Disco ball and activates it. An extension of this with the previous suggestion, `/Disco <player>` places the ball just above the target player and activates it. Since the audio for the ball seems to be played from the ball itself this should work.

- **Knife Damage Modifier**
v50 introduces the Butler enemy that drops a knife. Faster than shovel but less damage. Im sure you have already considered or implemented this. `/Knife <value>` just like /shovel changes the damage.

- **Haunt Player**
`/haunt <player>` Spawns and/or agros Ghost Girl on target. This could also be achieved by setting target's sanity to the lowest possible value.

- **Activate All Mechs/Old Birds**
`/Mech` activates all v50 Old Birds. This is the same effect as when the Apparatus item is pulled in vanilla gameplay.

- **Lightning Strike Target**
`/Strike <player>` Strikes target with lightning (would be funny to also work when conditions are normal on a moon).

- **Weapon Speed Modifier**
Extension of previous suggestion. I read on some LC forums clients can attack faster with the knife than the host. Allow the user to attack instantly upon clicking with knife/shovel or have a command that activates this `/rapidattack`, `/quickhit`, `/rapidhit`.

- **Hitman Command**
`/Hitman <player>` or some alternative of that can agro all butlers to one player or just send all butlers to go agro on the nearest player. Knives out!

- **Drug Player**
`/drug <player> <time value>` Forces the full effects TZP item onto a player for the specified time (mainly the vision blocking effects, speed effects will be funny in conjunction if visual effects cannot be isolated and applied to the target). Just like using the item normally, the effects slowly kick in and the target is confused and disoriented.

- **Force High Pitched Voice**
Extension of previous suggestion. Isolate voice changing effect of TZP and apply slowly (like normal use) to a target for the specified time so their voice becomes high pitched to everyone. `/Pitch <player> <time value>`.

Here are some QoL features/Suggestions:

- **Clarify use of some commands.**
Perhaps this is just me. I find the `/kill`, `/kill --enemy`, and `/fatality` difficult to use as its not clear how to use them in the README. Specifically, im not sure what "--" means. The ID/names of enemies is not clear (Do i type MouthDog? Dog? Eyeless Dog?). Is it possible to kill everyone by myself by typing `/kill -0` or `/kill -me` (assuming I am player 0)
`/fatality` is the same, I'll type `/fatality 1 Jester` and it will have an error and not work.

The `/end <player=-1>` command isn't clear to me how to use it. Does typing `/end` pull the ship lever? What impact does `<player=-1>` have on the result. Is it a criteria? Meaning does the command automatically end the game when 1 player is left if i type `/end 1`

`/grab` and `/build` also aren't clear to me. I'll type `/grab` sign and it will grab a key or bolt (I'm assuming it's because no sign exists to grab. I gotta ask incase I'm misunderstanding how its used). Same problem with `/build`. I typed `/build` cabinet and the table spawned.

Some clarification on how these commands are used (mainly what IDs need to be used for the desired outcome to occur) would be perfect.

- **Region Information**
When viewing a server or joining a server could some server info be provided in chat (ping speed, region, is host using LC Hax). An extension of this could be sorting servers by ping (although this is leaning towards mods like Sort by player count).

Those are all the suggestions I have. Thank you very much for your time and consideration.
If you'd like me to clarify anything please let me know.

**Comment by winstxnhdw** [comment]
Hey, thanks for taking the time to format these suggestions nicely. I see a few points that I am interested in adding. Unfortunately, it's currently exam month for me, so I won't have the time to push all of these out quickly. I'll be a lot more active again after the 9th of May.

To answer your questions:

> The /end <player=-1> command isn't clear to me how to use it. Does typing /end pull the ship lever? What impact does <player=-1> have on the result. Is it a criteria? Meaning does the command automatically end the game when 1 player is left if i type /end 1

Doing `/end <player=-1>` simply means that using `/end` without any arguments default to the player index of `-1`. Not sure if you realised, but when you use `/end`, it says that "player -1 has pulled the lever`. By specifying a player index or name, you can frame others for pulling the lever.

>  Is it possible to kill everyone by myself by typing /kill -0 or /kill -me (assuming I am player 0)

`/kill` without any arguments will kill yourself.

> The ID/names of enemies is not clear (Do i type MouthDog? Dog? Eyeless Dog?)

Don't think too much about it. Just type whatever makes sense. The `fatality` command uses fuzzy string matching, so it doesn't have to be accurate.

> /grab and /build also aren't clear to me. I'll type /grab sign and it will grab a key or bolt (I'm assuming it's because no sign exists to grab. I gotta ask incase I'm misunderstanding how its used). Same problem with /build. I typed /build cabinet and the table spawned.

Yes, you are right. It means no sign exists. `/build cabinet` should be `/build cupboard`.

**Reply by AlexOnGitHubb** [reply]
I formatted my post based on how I've seen you and other contributors respond to other questions. Figured it would make implementing any of them a bit easier. I'm glad you like it.

Please take your time. Exams are stressful. I wish you the best for your exams.
I imagine there is a coding, testing, bug fixing, and polishing/optimising phases to implementing features. I'm in no rush. Please take as long as you need to add any of the features to the quality you expect. 

Thank you so much for clarifying how the commands work! I didn't know i could frame other players! I will use that to only cause minor chaos amongst my friends.

I phrased the `/kill` question poorly. I keep killing myself. If i type `/kill -0` (im assuming -0 means everyone but player 0) will that kill everyone but me? Its purely the "--" in the README that has me confused, what does that represent? Keep in mind I have no coding knowledge (Im not sure what you mean my 'arguments').

I'll experiment more with the `fatality` command to get a better grip on how it works. I often just get error messages and nothing happens. Thank you for the insight!

Regarding the `/build` command, is there a resource I can use to get the names of the builds? If this uses the same fuzzy string matching it should be fine tho, I'll just need to experiment a bit more.

**Reply by winstxnhdw** [reply]
Commands have "arguments" and "flags".

`/kill alex` -> `alex` is the argument here
`/kill --all` -> `--all` is the flag here

As documented in the README, `kill` only has 2 flags currently; `--all` and `--enemy`.

> Regarding the /build command, is there a resource I can use to get the names of the builds? If this uses the same fuzzy string matching it should be fine tho, I'll just need to experiment a bit more.

The names should be similar to whatever you can find on the Lethal Company wikia; and yes, it does use fuzzy string matching.

It'll be helpful if you describe what's the error message you get for `fatality`.

**Reply by AlexOnGitHubb** [reply]
I understand now. Thank you for your insight.
I'll try out `Fatality `again and get back to you another day.
